md-proto/WebAssembly.md

```diff
@@ -0,0 +1,1884 @@
+this section does not cause validation for the whole module to fail and is
+instead treated as if the section was absent.
+
+The Names Section contains an [array] of function name descriptors, which each
+describe names for the function with the corresponding index in the module, and
+which contain:
+ - The function name, a [string].
+ - The names of the locals in the function, an [array] of [strings].
+
+> Name data is represented as an explicit section in the binary format, however
+in the text format it may be represented as an integrated part of the syntax for
+functions rather than as a discrete section.
+
+> The expectation is that, when a binary WebAssembly module is viewed in a
+browser or other development environment, the names in this section will be used
+as the names of functions and locals in the [text format](TextFormat.md).
+
+### Unknown Sections
+
+Modules may also contain unknown sections, with names other than those
+corresponding to known sections above. They have no semantic effect.
+
+### Function Bodies
+
+Function bodies contain an [array] of [types], which are declarations for
+locals, and a sequence of instructions.
+
+##### Positions within a function body
+
+A *position* within a function refers to an element of the sequence, or to an
+additional *return position* which is sequenced after the end of the function body.
+
+> In the binary encoding, positions are represented as byte offsets; in the text
+format, positions are represented with a special syntax.
+
+
+## Validation
+
+### Module Validation
+
+TODO: describe module validation
+
+### Function Validation
+
+For the duration of the validation of a function body, several data structures
+are created:
+ - A *control-flow stack*, with entries containing
+    - A [control-flow type](#control-flow-type).
+    - A *depth* value.
+ - A *[type] stack*.
+ - A *current position*.
+
+During validation, if either stack is empty when an element is to be popped from
+it, validation fails.
+
+#### Control Flow Type
+
+A *control-flow type* is either a [type](#types), `any`, which satisfies any
+type check, or `void`, which satisfies no check.
+
+> `any` is effectively serving as a kind of bottom type, and `void` is
+effectively serving as a kind of top type. Note that in some type theory
+contexts, it's common to use these names for exactly opposite purposes.
+
+#### Control Flow Type Merge
+
+To merge two [control-flow types]:
+ - If either control-flow type is `any`, the result is the other control flow
+   type.
+ - Otherwise, if the control-flow types are the same, the result is that control
+   flow type.
+ - Otherwise, validation fails with a type error.
+
+#### Function Validation Initialization
+
+The current position starts at the first position. The type stack begins empty.
+The control-flow stack starts with one entry. This entry's control-flow type is
+the first return type of the function, if it has any, or `void` otherwise. Its
+block depth is zero.
```

Comment:
What about the other return types of the function? Would this mean that a `br` could only return a single value, and that a `return` would be needed to return multiple-values?